File: exercise/product/src/app/service/customer.service.ts
// src/app/service/customer.service.ts

import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

// CRITICAL: Must import the Pirate interface from the correct file path
import { Pirate } from '../model/Pirate'; 

@Injectable({
  providedIn: 'root'
})
export class CustomerService {
  // Base endpoint for Pirate CRUD in your backend
  private apiUrl = 'http://localhost:8080/api/pirates'; 

  constructor(private http: HttpClient) { }
  
  /**
   * Handles user registration.
   * (Backend: POST to /api/pirates)
   */
  registerPirate(pirate: Pirate): Observable<Pirate> {
    return this.http.post<Pirate>(this.apiUrl, pirate);
  }

  /**
   * Handles user lookup for login verification.
   * (Backend: GET to /api/pirates/search/findByUsername?username=...)
   */
  loginPirate(username: string, password: string): Observable<Pirate> {
    // Send credentials as POST body (more secure than GET with query params)
    return this.http.post<Pirate>(`${this.apiUrl}/login`, {
      username: username,
      password: password
    });
  }

  /**
   * Retrieves a single Pirate by ID.
   * (Backend: GET to /api/pirates/{id})
   */
  findPirateById(id: number): Observable<Pirate> {
    return this.http.get<Pirate>(`${this.apiUrl}/${id}`);
  }

  /**
   * Deletes a Pirate entity. Required for the 'Delete Account' feature.
   * (Backend: DELETE to /api/pirates/{id})
   */
  deletePirate(id: number): Observable<void> {
    // Deletes the user record from the database
    return this.http.delete<void>(`${this.apiUrl}/${id}`); 
  }

  /**
   * Retrieves the full list of all Pirates (used by admin/GET endpoint).
   */
  findAllPirates(): Observable<Pirate[]> {
    return this.http.get<Pirate[]>(this.apiUrl);
  }
}

File: ecommerce/src/main/java/com/asis/service/CustomerService.java
package com.asis.service;

import com.asis.entity.Pirate;
import java.util.Optional;
import java.util.List;

public interface CustomerService {

    /**
     * Saves a new Pirate or updates an existing one.
     */
    Pirate savePirate(Pirate pirate);

    /**
     * Finds a Pirate by their unique ID.
     */
    Optional<Pirate> findPirateById(Long id);

    /**
     * Returns a list of all Pirates.
     */
    List<Pirate> findAllPirates();

    /**
     * Deletes a Pirate by ID.
     */
    void deletePirate(Long id);

    /**
     * Finds a Pirate by username (useful for login).
     */
    Optional<Pirate> findByUsername(String username);

    // The previous findPirateByUsername method is now named findByUsername to match the interface.\
    Pirate loginPirate(String username, String password);
}
